Library/Homebrew/unpack_strategy/directory.rb

```diff
@@ -16,33 +16,59 @@ def self.can_extract?(path)
+      ).void
+    }
+    def initialize(path, ref_type: nil, ref: nil, merge_xattrs: false, move: false)
+      super(path, ref_type:, ref:, merge_xattrs:)
+      @move = move
+    end
+
     private
 
     sig { override.params(unpack_dir: Pathname, basename: Pathname, verbose: T::Boolean).void }
     def extract_to_dir(unpack_dir, basename:, verbose:)
-      path_children = path.children
-      return if path_children.empty?
-
-      existing = unpack_dir.children
-
-      # We run a few cp attempts in the following order:
-      #
-      # 1. Start with `-al` to create hardlinks rather than copying files if the source and
-      #    target are on the same filesystem. On macOS, this is the only cp option that can
-      #    preserve hardlinks but it is only available since macOS 12.3 (file_cmds-353.100.22).
-      # 2. Try `-a` as GNU `cp -a` preserves hardlinks. macOS `cp -a` is identical to `cp -pR`.
-      # 3. Fall back on `-pR` to handle the case where GNU `cp -a` failed. This may happen if
-      #    installing into a filesystem that doesn't support hardlinks like an exFAT USB drive.
-      cp_arg_attempts = ["-a", "-pR"]
-      cp_arg_attempts.unshift("-al") if path.stat.dev == unpack_dir.stat.dev
-
-      cp_arg_attempts.each do |arg|
-        args = [arg, *path_children, unpack_dir]
-        must_succeed = print_stderr = (arg == cp_arg_attempts.last)
-        result = system_command("cp", args:, verbose:, must_succeed:, print_stderr:)
-        break if result.success?
-
-        FileUtils.rm_r(unpack_dir.children - existing)
+      move_to_dir(unpack_dir, verbose:) if @move
+
+      path.each_child do |child|
+        system_command! "cp",
+                        args:    ["-pR", (child.directory? && !child.symlink?) ? "#{child}/." : child,
+                                  unpack_dir/child.basename],
+                        verbose:
+      end
+    end
+
+    # Move files and non-conflicting directories from `path` to `unpack_dir`.
+    # Already existing subdirectories in `unpack_dir` are not modified.
+    #
+    # @raise [RuntimeError] on unsupported mv operation, e.g. overwriting a file with a directory
+    sig { params(unpack_dir: Pathname, verbose: T::Boolean).void }
+    def move_to_dir(unpack_dir, verbose:)
+      path.find(ignore_error: false) do |src|
+        next if src == path
+
+        dst = unpack_dir/src.relative_path_from(path)
+        if dst.exist?
+          dst_real_dir = dst.directory? && !dst.symlink?
+          src_real_dir = src.directory? && !src.symlink?
+          # Avoid moving a directory over an existing non-directory and vice versa.
+          # This outputs the same error message as GNU mv which is more user friendly than macOS mv.
+          raise "mv: cannot overwrite non-directory '#{dst}' with directory '#{src}'" if src_real_dir && !dst_real_dir
+          raise "mv: cannot overwrite directory '#{dst}' with non-directory '#{src}'" if !src_real_dir && dst_real_dir
```

Comment:
Will note that this is not fully backwards compatible.

Original `cp -pR` handling copies a child into a destination `child.basename`, which means that it will can copy a file into a directory of the same name, e.g. using commands directly
```console
❯ mkdir -p src dest/dir
❯ touch src/a src/b src/dir
❯ ls -1 src | xargs -I{} cp -pR src/{} dest/{}
❯ tree
.
├── dest
│   ├── a
│   ├── b
│   └── dir
│       └── dir
└── src
    ├── a
    ├── b
    └── dir
```

I don't think there is an exact scenario we do this, though it can probably be manually constructed by hacking up local `HOMEBREW_CELLAR/<formula>/...` path and then pouring a bottle.

Anyway, original behavior seemed odd so just kept error message for this case.